src/nspawn/nspawn.c

```diff
@@ -2594,9 +2594,15 @@ static int inner_child(
                 return -ESRCH;
         }
 
-        r = mount_systemd_cgroup_writable("", arg_unified_cgroup_hierarchy);
-        if (r < 0)
-                return r;
+	if (cg_ns_supported()) {
+		r = unshare(CLONE_NEWCGROUP);
+		if (r < 0)
```

Comment:
On Sat, Jun 25, 2016 at 03:08:20AM -0700, Evgeny Vereshchagin wrote:
> > @@ -2594,9 +2594,15 @@ static int inner_child(
> >                  return -ESRCH;
> >          }
> >  
> > -        r = mount_systemd_cgroup_writable("", arg_unified_cgroup_hierarchy);
> > -        if (r < 0)
> > -                return r;
> > +	if (cg_ns_supported()) {
> > +		r = unshare(CLONE_NEWCGROUP);
> > +		if (r < 0)
> 
> >  This is why they do not appear in the container which checks the available controllers in the v1 hierarchy.
> 
> But why do we need to check the `v1`-controllers on the `v2`-hierarchy?
> 
> >  In the case of cgroup namespaces the container will be able to mount a cgroup filesystem by itself just as on normal system bootup. 
> 
> Yeah. But we shouldn't mount `v1` on `v2` (and vice versa)
> 
> `master`:
> ```sh
> -bash-4.3# systemd-nspawn -D /nspawn-root -b 3
> ...
> container# grep cgroup /proc/self/mounts
> cgroup /sys/fs/cgroup cgroup2 ro,nosuid,nodev,noexec,relatime 0 0
> cgroup /sys/fs/cgroup/machine.slice/machine-nspawn\134x2droot.scope cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
> 
> container# cat /proc/1/cgroup
> 0::/machine.slice/machine-nspawn\x2droot.scope/init.scope
> ```
I can reproduce this behavior with systemd master independent of this patch.
Sorry, I'm a little confused as to what you're getting at here.